src/Features/Core/Portable/Shared/TestHooks/IAsynchronousOperationListenerProvider.cs

```diff
@@ -0,0 +1,209 @@
+        {
+            return (IAsynchronousOperationWaiter)GetListener(featureName);
+        }
+
+        /// <summary>
+        /// Wait for all of the <see cref="IAsynchronousOperationWaiter"/> instances to finish their
+        /// work.
+        /// </summary>
+        /// <remarks>
+        /// This is a very handy method for debugging hangs in the unit test.  Set a break point in the 
+        /// loop, dig into the waiters and see all of the active <see cref="IAsyncToken"/> values 
+        /// representing the remaining work.
+        /// </remarks>
+        public async Task WaitAllAsync(string[] featureNames = null, Action eventProcessingAction = null)
+        {
+            var smallTimeout = TimeSpan.FromMilliseconds(10);
+
+            Task[] tasks = null;
+            while (true)
+            {
+                var waiters = GetCandidateWaiters(featureNames);
+                tasks = waiters.Select(x => x.CreateWaitTask()).Where(t => !t.IsCompleted).ToArray();
+
+                if (tasks.Length == 0)
+                {
+                    // no new pending tasks
+                    break;
+                }
+
+                do
+                {
+                    // wait for all current tasks to be done for the time given
+                    if (Task.WaitAll(tasks, smallTimeout))
+                    {
+                        // current set of tasks are done.
+                        // see whether there are new tasks added while we were waiting
+                        break;
+                    }
+
+                    // certain test requires some event queues to be processed
+                    // for waiter tasks to finish such as Dispatcher queue
+                    eventProcessingAction?.Invoke();
+
+                    // in unit test where it uses fake foreground task scheduler such as StaTaskScheduler
+                    // we need to yield for the scheduler to run inlined tasks
+                    await Task.Yield();
+                } while (true);
+            }
+
+            foreach (var task in tasks)
+            {
+                if (task.Exception != null)
+                {
+                    throw task.Exception;
+                }
+            }
+        }
+
+        public bool HasPendingWaiter(params string[] featureNames)
+        {
+            var waiters = GetCandidateWaiters(featureNames);
+            return waiters.Any(w => w.HasPendingWork);
+        }
+
+        /// <summary>
+        /// Get all saved DiagnosticAsyncToken to investigate tests failure easier
+        /// </summary>
+        public List<AsynchronousOperationListener.DiagnosticAsyncToken> GetTokens()
+        {
+            return _singletonListeners.Values.Cast<AsynchronousOperationListener>().Where(l => l.TrackActiveTokens).SelectMany(l => l.ActiveDiagnosticTokens).ToList();
+        }
+
+        private static bool IsEnabled
+        {
+            get
+            {
+                if (!s_enabled.HasValue)
```

Comment:
didn't use lock for this and the one below. these are not expected to called multiple time. but only once for 1 VS session.